### PR TITLE
Add support for parsing attributes and other improvements

### LINKED ISSUE #22: Finish attribute value parsing

### PR DESCRIPTION
**Pull request by kennykerr**
This update adds the ability to parse basic built-in and custom WinRT attributes, the last remaining feature required by WinRT. It isn't complete, but it does cover the bases. What remains is substantially more testing, which I have started addressing here with some more testing infrastructure that will need to be fleshed out considerably. Notables:

* Got rid of the "not_used" marker variants. The result is much simpler to read and doesn't confuse IntelliSense.
* Removed a lot of dead metadata tables not used by WinRT.
* Simplified error handling
* Renamed Database -> File to avoid giving the wrong impression as the Database on its own not a complete representation and needs the Reader
* The row types can now follow the breadcrumbs back to the Reader. This is key to supporting more complex queries that require type lookup (e.g. attributes)

Finally, I now include a test IDL file along with a checked-in winmd file for testing. The latter is checked in to avoid requiring MIDL as part of the build. Once the winmd parser can also produce winmd files, this will no longer be required, but until then it helps to test various edge cases without relying on recent OS metadata.

Fixes #22

### CODE REVIEW

File: src/signatures.rs [hunk 1]
```diff
@@ -202,57 +344,94 @@ fn read_expected(bytes: &mut &[u8], expected: u32) -> ParseResult<bool> {
...
+    std::str::from_utf8(string_bytes).unwrap()
+}
+
+fn read_i8(bytes: &mut &[u8]) -> i8 {
+    let (value_bytes, rest) = bytes.split_at(std::mem::size_of::<i8>());
+    *bytes = rest;
+    i8::from_le_bytes(value_bytes.try_into().unwrap())
+}
+
+fn read_u8(bytes: &mut &[u8]) -> u8 {
```
Comment (kennykerr) [inline src/signatures.rs]: Not too happy with these - seems very repetitive, but couldn't figure out how to write this as a generic function - much like a function template in C++.

File: macros/src/lib.rs [hunk 2]
```diff
@@ -19,27 +19,35 @@ pub fn type_code(args: TokenStream, input: TokenStream) -> TokenStream {
...
-    for (index, variant) in input.variants.iter().enumerate() {
+    for variant in input.variants.iter() {
         let camel = &variant.ident;
         let camel_name = camel.to_string();
 
-        if camel_name != "not_used" {
-            let snake = syn::Ident::new(&to_snake(&camel_name), camel.span());
-            let index = index as u32;
+        if let Some((_, value)) = &variant.discriminant {
+            if let syn::Expr::Lit(value) = value {
```
Comment (rylev) [inline macros/src/lib.rs]: You can inline this `if let` into the one above:

`if let Some((_, syn::Expr::Lit(value))) = &variant.discriminant {`

File: src/helpers.rs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+use crate::error::*;
+
+pub fn split_type_name(name: &str) -> ParseResult<(&str, &str)> {
+    let index = name.rfind('.').ok_or_else(|| ParseError::InvalidTypeName)?;
+    Ok((name.get(0..index).unwrap(), name.get(index + 1..).unwrap()))
```
Comment (rylev) [inline src/helpers.rs]: nit: if you're just going to `unwrap` no need to use `get`. You can do `name[0..index]`

File: src/reader.rs [hunk 4]
```diff
@@ -1,31 +1,27 @@
-use crate::database::*;
 use crate::error::*;
+use crate::file::*;
+use crate::helpers::*;
 use crate::tables::*;
 
+pub struct Reader {
+    files: std::vec::Vec<File>,
```
Comment (rylev) [inline src/reader.rs]: nit: `Vec` doesn't need to be namespaced

File: src/reader.rs [hunk 5]
```diff
@@ -38,10 +34,20 @@ struct NamespaceData {
     delegates: std::vec::Vec<(u32, u32)>,
 }
 
-pub struct Namespace<'a> {
-    reader: &'a Reader,
-    name: &'a str,
-    types: &'a NamespaceData,
+impl<'a> Iterator for NamespaceIterator<'a> {
+    type Item = Namespace<'a>;
+    fn next(&mut self) -> Option<Self::Item> {
```
Comment (rylev) [inline src/reader.rs]: This iterator seems like it can be accomplished using `map` on iterator. Is there a need for this custom iterator?

File: src/signatures.rs (hunk 1 above)
Comment (rylev) [reply src/signatures.rs]: You could use a macro to write these.

File: src/reader.rs (hunk 5 above)
Comment (kennykerr) [reply src/reader.rs]: I'll check.

File: src/helpers.rs (hunk 3 above)
Comment (kennykerr) [reply src/helpers.rs]: Right, forgot about that. 👍
